api.lua

```diff
@@ -0,0 +1,53 @@
+
+function wrench.register_node(name, def)
+	assert(type(name) == "string", "wrench.register_node invalid type for name")
+	assert(type(def) == "table", "wrench.register_node invalid type for def")
+	local node_def = minetest.registered_nodes[name]
+	if node_def then
+		def = table.copy(def)
+		if def.drop == true and type(node_def.drop) == "string" then
+			def.drop = node_def.drop
+		elseif def.drop and type(def.drop) ~= "string" then
+			minetest.log("warning", "[wrench] Ignoring invalid type for drop in definition for "..name)
+			def.drop = nil
+		end
+		if def.drop and not wrench.registered_nodes[def.drop] then
+			if def.drop ~= name then
+				minetest.log("warning", "[wrench] Ignoring unsupported node for drop in definition for "..name)
+			end
+			def.drop = nil
+		end
+		wrench.registered_nodes[name] = def
+	else
+		minetest.log("warning", "[wrench] Attempt to register unknown node for wrench: "..name)
+	end
+end
+
+function wrench.blacklist_item(name)
+	assert(type(name) == "string", "wrench.blacklist_item invalid type for name")
+	local node_def = minetest.registered_items[name]
+	if node_def then
+		wrench.blacklisted_items[name] = true
+	else
+		minetest.log("warning", "[wrench] Attempt to blacklist unknown item for wrench: "..name)
+	end
+end
+
+minetest.register_on_mods_loaded(function()
+	for name, def in pairs(wrench.registered_nodes) do
+		local node_def = minetest.registered_nodes[name]
+		if node_def then
+			local old_after_place = node_def.after_place_node
+			minetest.override_item(name, {
+				after_place_node = function(...)
+					if not wrench.restore_node(...) and old_after_place then
+						return old_after_place(...)
+					end
+				end
+			})
+		else
+			minetest.log("warning", "[wrench] Registered node is now unknown: "..name)
+			wrench.registered_nodes[name] = nil
+		end
+	end
+end)
```

Comment:
I'd recommend reverting this change to registration API unless there's very good reason to execute it only after every other mod is loaded (remember that all dependencies are loaded already).

It does:
* prevent registering nodes runtime using api
* makes override ordering a lot harder to predict

Reply:
It does three things:

- Ensures the override only happens once, even if a node is registered multiple times.
- Ensures that it's the last override, regardless of what mods override the callback.
- Allows unregistering wrench support for a node, before the override happens.

I don't know of any other way to do that.

Reply:
> Ensures the override only happens once, even if a node is registered multiple times.

Doing override multiple times does not matter, it is same as unregistering and registering again. Could just override previous override, should not be anything wrong with that.

> Ensures that it's the last override, regardless of what mods override the callback.

No, it does not. It just makes it more likely that it is last override done from said callback for mods that do not depend on wrench. On the other hands for mods that do depend on wrench... well you know already.

> Allows unregistering wrench support for a node, before the override happens.

This is nice thing (is there actual use cases?). However having it available runtime can allow unregistering wrench support any time. Maybe think for a while if there would be a better way to provide this functionality.

Real override unregister/rollback can be done, that should not be problem but this being one of reasons why registration is delayed and made all at once begs questions: how and why is unregistering used currently?

> I don't know of any other way to do that.

Not sure if I know how to do that but it is very hard to imagine that delayed registration is really needed, provided points at least are not really supporting it but also I've not worked on this so do not know if there actually is something that makes delayed registration work better.

Reply:
I tried some other ways of implementing this (caching the original callbacks, checking if the node had been overridden), but I couldn't find a way to cleanly implement everything; delaying the registration is the best solution I know of.

As an example, imagine some mods A, B, C and D, with immediate registration: 
- A registers a node.
- B registers wrench support for the node from mod A. (override happens, wrapping `after_place_node`)
- C, unaware of mod B, also adds wrench support for the node. (`after_place_node` is now double wrapped)
- D, unaware of B and C, overrides the node from mod A. (`after_place_node` gets replaced, no wrapping for wrench functionality)

If the override (wrapping of `after_place_node`) gets delayed until all mods have loaded, then neither of the two problems happen.

Reply:
I think there's no reason to account for _"B registers wrench support for the node from mod A"_, in that case A and B should be well aware of each other if it is modpack, server customization mod or game. If it is just basic mod then it should not be doing that (if it handles stuff in A it should already know what A already does).

For **any** double wrapping it should be possible to simply check if node override is already done or not.

But for _"D, unaware of B and C, overrides the node from mod A."_:
This probably could be simpler put as _"D overrides after_place_node from mod A which registered node for wrench"_ which is scenario where D should know about wrench but probably wont take wrench internal mechanisms into account (and should not have to worry about it).
That causes situation where override has to be delayed (without having to consider B or C at all)?
For this there probably is no good solutions around it without delaying registration.

Reply:
Well, only solution for that would be to drop API from wrench and only registering support internally but that would mean no API and therefore nobody else being able to add wrench support for nodes, that does not sound like good solution (it has upsides but would deny too much in my opinion).

Reply:
"*A registers a node.*" means `minetest.register_node()`, not registering for wrench ;)

Reply:
> "_A registers a node._" means `minetest.register_node()`, not registering for wrench ;)

Well, then you only have to consider B and D and instead of A and D. Other than that scenario seems to stay unaffected.